semgrep-core/src/core/ast/Ugly_print_AST.ml

```diff
@@ -0,0 +1,122 @@
+ * If UglyPrinter encounters a construct that it is unable to handle, it returns
+ * `None`. It is crucial that maintainers preserve this property, since
+ * otherwise it is possible that we could present syntactically-incorrect code
+ * as part of, for example, an autofix.
+ *
+ * Classes are used here so that we can take advantage of inheritance for
+ * implementation-sharing and dynamic dispatch. Dynamic dispatch allows for the
+ * hybrid printer used for autofix, where we use the original text from the
+ * target or pattern for nodes that have been lifted unchanged. This is
+ * implemented by inheriting from the plain printer. With a different
+ * architecture, printer authors would need to deal with the logic for this.
+ * This way, printer authors need only focus on printing AST correctly.
+ *
+ * OCaml's powerful object system will also allow us to share implementation
+ * between printers to whatever extent is appropriate. `Pretty_print_AST`, for
+ * example, gates on language within each function. We could replicate that here
+ * if we like.
+ *
+ * This module should replace `Pretty_print_AST`. The main motivator for
+ * creating a new printer module is that the use of classes allows for the
+ * hybrid printer in autofix, as described above. `Pretty_print_AST` also does
+ * not consistently report when it is unable to print a node correctly. We can
+ * address that as we migrate it here.
+ *
+ * [1] https://homepages.inf.ed.ac.uk/wadler/papers/prettier/prettier.pdf *)
+(******************************************************************************)
+
+class type printer_t =
+  object
+    method print_any : G.any -> Immutable_buffer.t option
+    method print_expr : G.expr -> Immutable_buffer.t option
+    method print_expr_kind : G.expr_kind -> Immutable_buffer.t option
+    method print_argument : G.argument -> Immutable_buffer.t option
+    method print_arguments : G.arguments -> Immutable_buffer.t option
+
+    (* TODO Add more nodes as needed. *)
+  end
+
+(* Some if and only if `f` returns `Some` for all elements. Short circuits when
+ * the first `None` is encountered.
+ *
+ * TODO Make tail recursive?
+ * TODO Move to Common?
+ *)
+let rec map_all f = function
+  | [] -> Some []
+  | hd :: tl ->
+      let* hd = f hd in
+      let* tl = map_all f tl in
+      Some (hd :: tl)
+
+(* We could consider partially replacing this with an autogenerated reduce
+ * visitor, where `plus` simply returns None. That could cut down on the
+ * boilerplate of handling e.g. the any case, but since every node needs to be
+ * handled explicitly for complete pretty-printing, it might be more trouble
+ * than it's worth. *)
+class base_printer : printer_t =
+  object (self)
+    method print_any =
+      function
+      | G.E expr -> self#print_expr expr
+      | G.Ar arg -> self#print_argument arg
+      (* TODO Handle other kinds *)
+      | _ -> None
+
+    method print_expr { e; _ } = self#print_expr_kind e
+    method print_expr_kind _ = None
```

Comment:
Just skimming this, not doing a full review, but maybe a Result type would be more clear? It's not immediately obvious to me that "None" means "failure"

Reply:
I don't feel strongly, but I do have a preference for `option` here. `option` is frequently used in the standard library and elsewhere in cases where, for whatever reason, a function might not produce a value. `Result.t` is a little bit more annoying to deal with, but it's worth it when you need to attach some data explaining why you don't have a value. I don't think we need such data here. With that said, I'm happy to make the change if you feel strongly.

Reply:
Yeah, that's a good point. Using `Result.t` would let us report with granularity where exactly the printing failed. I agree that that's valuable, but I think doing that well would be a large change to this PR. How about for this PR, I add logging so that we can at least tell which large phase failed, and then experiment with `Result.t` later this week as a followup?